### PR TITLE
tests: cpp: cxx: include top-level driver headers

### PR DESCRIPTION
**Pull request by henrikbrixandersen**
Include (almost) all top-level driver header files to ensure error-free C++ compilation. This revealed a few issues, which are fixed by this PR as well.

### CODE REVIEW

File: include/drivers/dma.h [hunk 1]
```diff
@@ -479,7 +479,7 @@ static inline int z_impl_dma_request_channel(const struct device *dev,
 	const struct dma_driver_api *api =
 		(const struct dma_driver_api *)dev->api;
 	/* dma_context shall be the first one in dev data */
-	struct dma_context *dma_ctx = dev->data;
+	struct dma_context *dma_ctx = (struct dma_context *)dev->data;
```
Comment (stephanosio) [inline include/drivers/dma.h:482]: It is a pity that C++ requires this ... nothing we can do about it.

**Review by stephanosio — DISMISSED** [review]

**Review by teburd — DISMISSED** [review]

**Review by stephanosio — DISMISSED** [review]

**Comment by henrikbrixandersen** [comment]
Rebased after 53ef68d4598b2f9005c5da3fc0b860ca1999d350.

**Review by stephanosio — APPROVED** [review]

File: tests/subsys/cpp/cxx/src/main.cpp [hunk 2]
```diff
@@ -17,16 +17,62 @@
 #include <device.h>
 #include <kernel.h>
 #include <net/buf.h>
-#include <sys/byteorder.h>
+/* #include <sys/byteorder.h> conflicts with __bswapXX on native_posix */
```
Comment (cfriedt) [inline tests/subsys/cpp/cxx/src/main.cpp:20]: Do we have a separate issue for this, for tracking?

**Review by cfriedt — APPROVED** [review]

File: tests/subsys/cpp/cxx/src/main.cpp (hunk 2 above)
Comment (henrikbrixandersen) [reply tests/subsys/cpp/cxx/src/main.cpp:20]: @cfriedt See https://github.com/zephyrproject-rtos/zephyr/issues/44324#issuecomment-1096271561

**Pull request merged**